File: code/algomatrix/Q2 Data Structures Correction(Please Ignore Others)/Group 2/heaps.php
<?php
class MinHeap {
    protected $heap;
    protected $size;

    public function __construct() {
        $this->heap = array();
        $this->size = 0;
    }

    protected function getParentIndex($index) {
        return ($index - 1) / 2;
    }

    protected function getLeftChildIndex($index) {
        return 2 * $index + 1;
    }

    protected function getRightChildIndex($index) {
        return 2 * $index + 2;
    }

    protected function hasParent($index) {
        return $this->getParentIndex($index) >= 0;
    }

    protected function hasLeftChild($index) {
        return $this->getLeftChildIndex($index) < $this->size;
    }

    protected function hasRightChild($index) {
        return $this->getRightChildIndex($index) < $this->size;
    }

    protected function swap($index1, $index2) {
        $temp = $this->heap[$index1];
        $this->heap[$index1] = $this->heap[$index2];
        $this->heap[$index2] = $temp;
    }

    public function insert($value) {
        $this->heap[$this->size] = $value;
        $this->size++;
        $this->heapifyUp();
    }

    protected function heapifyUp() {
        $index = $this->size - 1;
        while ($this->hasParent($index) && $this->heap[$this->getParentIndex($index)] > $this->heap[$index]) {
            $this->swap($this->getParentIndex($index), $index);
            $index = $this->getParentIndex($index);
        }
    }

    public function extractMin() {
        if ($this->size == 0) {
            return null;
        }
        $minValue = $this->heap[0];
        $this->heap[0] = $this->heap[$this->size - 1];
        $this->size--;
        $this->heapifyDown();
        return $minValue;
    }

    protected function heapifyDown() {
        $index = 0;
        while ($this->hasLeftChild($index)) {
            $smallerChildIndex = $this->getLeftChildIndex($index);
            if ($this->hasRightChild($index) && $this->heap[$this->getRightChildIndex($index)] < $this->heap[$smallerChildIndex]) {
                $smallerChildIndex = $this->getRightChildIndex($index);
            }
            if ($this->heap[$index] < $this->heap[$smallerChildIndex]) {
                break;
            } else {
                $this->swap($index, $smallerChildIndex);
            }
            $index = $smallerChildIndex;
        }
    }

    public function display() {
        echo implode(', ', $this->heap) . "\n";
    }
}

File: code/algomatrix/Q2 Data Structures Correction(Please Ignore Others)/Group 3/heaps.java
public class MinHeap {
    private int[] heap;
    private int size;
    private int capacity;

    public MinHeap(int capacity) {
        this.capacity = capacity;
        this.size = 0;
        this.heap = new int[capacity];
    }

    private int getParentIndex(int index) {
        return (index - 1) / 2;
    }

    private int getLeftChildIndex(int index) {
        return 2 * index + 1;
    }

    private int getRightChildIndex(int index) {
        return 2 * index + 2;
    }

    private boolean hasParent(int index) {
        return getParentIndex(index) >= 0;
    }

    private boolean hasLeftChild(int index) {
        return getLeftChildIndex(index) < size;
    }

    private boolean hasRightChild(int index) {
        return getRightChildIndex(index) < size;
    }

    private void swap(int index1, int index2) {
        int temp = heap[index1];
        heap[index1] = heap[index2];
        heap[index2] = temp;
    }

    public void insert(int value) {
        if (size == capacity) {
            throw new IllegalStateException("Heap is full");
        }
        heap[size] = value;
        size++;
        heapifyUp();
    }

    public int extractMin() {
        if (size == 0) {
            throw new IllegalStateException("Heap is empty");
        }
        int minValue = heap[0];
        heap[0] = heap[size - 1];
        size--;
        heapifyDown();
        return minValue;
    }

    private void heapifyUp() {
        int index = size - 1;
        while (hasParent(index) && heap[getParentIndex(index)] > heap[index]) {
            swap(getParentIndex(index), index);
            index = getParentIndex(index);
        }
    }

    private void heapifyDown() {
        int index = 0;
        while (hasLeftChild(index)) {
            int smallerChildIndex = getLeftChildIndex(index);
            if (hasRightChild(index) && heap[getRightChildIndex(index)] < heap[smallerChildIndex]) {
                smallerChildIndex = getRightChildIndex(index);
            }
            if (heap[index] < heap[smallerChildIndex]) {
                break;
            } else {
                swap(index, smallerChildIndex);
            }
            index = smallerChildIndex;
        }
    }

    public void display() {
        for (int i = 0; i < size; i++) {
            System.out.print(heap[i] + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        MinHeap minHeap = new MinHeap(10);
        minHeap.insert(5);
        minHeap.insert(10);
        minHeap.insert(3);
        minHeap.insert(7);
        minHeap.insert(1);

        minHeap.display(); // Output: 1 3 5 10 7

        System.out.println("Extracted min: " + minHeap.extractMin()); // Output: Extracted min: 1
        minHeap.display(); // Output: 3 7 5 10
    }
}